### PR TITLE
Improve bucketed negated spine

### PR DESCRIPTION
**Pull request by vmarcos**
This PR performs improvements to the spine types in selected type-based dispatch calls in reduce. Most improvements are concentrated in bucketed reductions, where only the first level of a hierarchy of arrangements is left at `RowValSpine`, while all remaining levels can use `RowRowSpine`. Additionally, a refactoring envisioning reuse is done on
merges and splits of `(key, hash)` pairs.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

The first commit is from #24375 and can be ignored. Only the fourth commit is relevant here.

You probably want to hide whitespace when reviewing, as `cargo fmt` changed  indentation of a code block.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/render/reduce.rs [hunk 1]
```diff
@@ -2125,6 +2189,63 @@ impl Columnation for Accum {
...
+    where
+        I: Iterator<Item = Datum<'a>>,
+    {
+        let binding = SharedRow::get();
+        let mut row_builder = binding.borrow_mut();
+        let mut row_packer = row_builder.packer();
+        row_packer.push(Datum::UInt64(val));
+        row_packer.extend(key);
+        row_builder.clone()
+    }
```
Comment (antiguru) [inline src/compute/src/render/reduce.rs:2223]: You might be able to combine this into one function that takes two iterators that yield datums. Not sure, but worth a try.

File: src/compute/src/render/reduce.rs [hunk 2]
```diff
@@ -1381,15 +1447,13 @@ where
                     row_builder.packer().push(value);
                     (key, row_builder.clone())
                 })
-                .map(|k| (k, ()))
-                .mz_arrange::<KeySpine<(Row, Row), _, _>>(
-                    "Arranged Accumulable Distinct [val: empty]",
-                )
-                .mz_reduce_abelian::<_, KeySpine<_, _, _>>(
+                .map(move |(k, v)| (key_val_pairer.merge(k, v), ()))
```
Comment (antiguru) [inline src/compute/src/render/reduce.rs:1450]: Can this `map` be inlined into the previous `map`?

File: .config/shepherdly.yml
[the remark this answers is not in the document]
Comment (vmarcos) [reply .config/shepherdly.yml:1]: This got invoked as a result of an out-of-order rebase, so no YML here to be reviewed.

File: src/compute/src/render/reduce.rs (hunk 2 above)
Comment (vmarcos) [reply src/compute/src/render/reduce.rs]: Done in https://github.com/MaterializeInc/materialize/pull/24375, rebased here now.

File: src/compute/src/render/reduce.rs [hunk 3]
```diff
@@ -874,21 +886,21 @@ where
...
     ) -> Arranged<S, TraceAgent<Tr>>
     where
         S: Scope<Timestamp = G::Timestamp>,
         Tr::ValOwned: MaybeValidatingRow<(), String>,
         Tr: Trace
             + for<'a> TraceReader<
-                Key<'a> = &'a (Row, Row),
-                KeyOwned = (Row, Row),
+                Key<'a> = DatumSeq<'a>,
+                KeyOwned = Row,
```
Comment (antiguru) [inline src/compute/src/render/reduce.rs:898]: `KeyOwned` doesn't need to be specified I think!

File: src/compute/src/render/reduce.rs [hunk 4]
```diff
@@ -1152,75 +1181,77 @@ where
...
-    ) -> Collection<S, (Row, R), Diff>
+    ) -> Arranged<S, TraceAgent<Tr>>
     where
         S: Scope<Timestamp = G::Timestamp>,
-        R: MaybeValidatingRow<Row, Row>,
+        Tr::ValOwned: MaybeValidatingRow<Row, Row>,
+        Tr: Trace
+            + for<'a> TraceReader<
+                Key<'a> = DatumSeq<'a>,
+                KeyOwned = Row,
```
Comment (antiguru) [inline src/compute/src/render/reduce.rs:1195]: Same here: `KeyOwned` doesn't need to be specified.

**Review by antiguru — APPROVED** [review]
Looks good, thank you! I left some minor comments inline, and I assume you'll rebase this PR before merging, so I didn't repeat my comments on the `Pairer`.

File: src/compute/src/render/reduce.rs (hunk 4 above)
Comment (vmarcos) [reply src/compute/src/render/reduce.rs]: Fixed here now!

**Comment by vmarcos** [comment]
TFTR!

**Pull request merged**